Improve combine node 2

fakob opened the pull request:
This is a temporary PR, as I am still looking for the issue with the last item missing the override. Maybe you can have a look through if you see something obviously wrong?

magnificus commented on src/nodes/draw/draw.tsx:
```diff
@@ -476,15 +610,11 @@ export class DRAW_COMBINE_ARRAY extends DRAW_Interactive_Base {
-          i,
-          executions,
-        );
+        addShallowContainerEventListeners(shallowContainer, this, i);
```
missing executions here, intended???

fakob replied:
executions is not used in the eventlistener, so I thought that I can remove it completely

magnificus replied:
OK

magnificus commented on src/nodes/draw/draw.tsx:
```diff
@@ -344,14 +398,93 @@ export class DRAW_Combine extends DRAW_Base {
       positionKeys.reverse();
     }
 
+    const margin = {
```
fairly certain something with this logic here is quacky, it looks very complex, im sure we can simplify it, this will likely also fix the bug

magnificus commented on src/nodes/draw/abstract.tsx:
```diff
@@ -352,12 +414,29 @@ export abstract class DRAW_Base extends PPNode {
-        myContainerBounds.height + margin.top + margin.bottom,
-      );
+
+      let bgX = -margin.left + myContainerBounds.x;
```
this does not look good I think

fakob commented:
I have tried to keep the needed complexity completely within the combine node. Catering for this many options was challenging, but I think it is worth it.
* Updated the combine node and added the ability to
  * layout/position the inputs
    * manually (manual) - old behaviour
    * automatically separating them with a gap (gap)
    * automatically spreading them out between a fixed distance (spread)
  * set the direction of the layout (vertical or horizontal)
  * set the container size
    * based on the content size (hug)
    * based on the parent size (fill)
    * based on a fixed size (fixed)
  * align the inputs inside the container left/center/right, top/center/bottom
* Cleaned up abstract class
  * removed pivot (default pivot is now centered)
  * moved margin and background to combine node

<img width="937" alt="Screenshot 2024-03-29 at 16 32 00" src="https://github.com/fakob/plug-and-play/assets/4619772/1fc3280a-a654-412f-96c6-6356df06e1b6">

magnificus commented on src/nodes/draw/abstract.tsx:
```diff
@@ -275,13 +266,27 @@ export abstract class DRAW_Base extends PPNode {
-  private handleDrawing(drawingFunction: any): void {
+  private handleDrawing(drawingFunction: any, inputObject): void {
+    let passedInOverrideSettings;
+    // example for passing in overrides
```
can remove

fakob replied:
will do

magnificus commented on src/nodes/draw/abstract.tsx:
```diff
@@ -275,13 +266,27 @@ export abstract class DRAW_Base extends PPNode {
   });
 
-  private handleDrawing(drawingFunction: any): void {
+  private handleDrawing(drawingFunction: any, inputObject): void {
```
inputobject not used? if so lets remove

fakob replied:
will do

magnificus commented on src/nodes/draw/draw.tsx:
```diff
@@ -344,16 +450,261 @@ export class DRAW_Combine extends DRAW_Base {
+            spacingValue * (graphicsKeys.length - 1);
+        break;
+      case widthHeightFillOptions[1].text: // fill
+        if (parentBgWidth) {
```
avoid nullcheck here, we should be able to say with confidence that it exists

fakob replied:
Well, the point is that `parentBgWidth` is only passed down/relevant when the parent is another combine node. However, the parent can be a multiplier node which does not pass this property down. In this case it just falls back to use the fixed with.

magnificus commented on src/nodes/draw/draw.tsx:
```diff
@@ -344,16 +450,261 @@ export class DRAW_Combine extends DRAW_Base {
-        const position = new PIXI.Point(
-          inputObject[positionKey].x,
-          inputObject[positionKey].y,
+    const bgColor = parseValueAndAttachWarnings(
```
this function looks super verbose, can we simplify it? We should at least be able to split this function into several smaller ones.

fakob replied:
I had tried for some time to make it simpler, but due to its diverse functionality, I was not able to do that. I have now split certain parts into separate functions.

magnificus reviewed (changes requested):
Looks promising and will maybe tie in well with the performance improvement (?) but left some comments

magnificus commented on src/nodes/draw/draw.tsx:
```diff
@@ -344,16 +572,203 @@ export class DRAW_Combine extends DRAW_Base {
+          case horizontalAlignmentOptions[1].text: // center
+            currentPositionX =
+              padding.left +
+              (predefinedBackgroundWidth -
```
same as comment above

magnificus commented on src/nodes/draw/draw.tsx:
```diff
@@ -344,16 +572,203 @@ export class DRAW_Combine extends DRAW_Base {
+              break;
+            case verticalAlignmentOptions[2].text: // bottom
+              currentPositionY =
+                padding.top + predefinedBackgroundHeight - bounds.height / 2;
```
one idea, is it possible to turn bounds or predefinedBackgroundHeight into something like a "heightToUse" varialble? then the difference could happen up there and dont need to diverge paths here

magnificus commented on cypress/e2e/pnp/parallelizable/testCombineObjects.cy.ts:
```diff
@@ -0,0 +1,464 @@
+import { doWithTestController, openNewGraph } from '../helpers';
```
good with test

magnificus commented on cypress/e2e/pnp/parallelizable/testCombineObjects.cy.ts:
```diff
@@ -0,0 +1,464 @@
+      );
+      testController.executeNodeByID('DRAW_Combine-3');
+    });
+    cy.wait(100);
```
are these waits needed? cant we use the await toast thing instead?

fakob replied:
I believe that I had issues when I ran `doWithTestController` back to back, but I can check again.

fakob replied:
So it seems that I do need a short wait after every`executeNodeByID` call.

magnificus commented on cypress/e2e/pnp/parallelizable/testCombineObjects.cy.ts:
```diff
@@ -0,0 +1,464 @@
+      );
+      testController.executeNodeByID('DRAW_Combine-3');
+    });
+    cy.wait(100);
```
same as above and below etc

magnificus commented on src/TestController.ts:
```diff
@@ -50,6 +51,20 @@ export default class TestController {
+      child.x,
+      child.y,
+    ]);
+    console.log(coordinates);
```
remove log

magnificus commented on src/components/ContextMenus.tsx:
```diff
@@ -610,17 +614,28 @@ export const NodeContextMenu = (props) => {
-          </MenuItem>
+        {areDrawNodesSelected && (
+          <>
+            <MenuItem
```
do you want the extract macro thing to be related to this draw node specific thing? Possible typo

fakob replied:
Good catch! That is a mistake.

magnificus commented on src/nodes/draw/draw.tsx:
```diff
@@ -319,20 +319,252 @@ export class DRAW_Text extends DRAW_Base {
+  return graphicsKeys.reduce((acc, graphicsKey, index) => {
+    const drawingFunction: DeferredPixiTypeInterface = inputObject[graphicsKey];
+    const newHash = uuid();
+    drawingFunction.hash = () => newHash;
```
drawingFunction.hash = uuid

fakob replied:
Ah yes.

fakob replied:
Hmm, actually that ended up in an error inside DRAW_Get_Bounds.getDrawingBounds
> TypeError: Cannot read properties of undefined (reading '×')
As I only need to set a unique value once. I kept it.

magnificus commented on src/nodes/draw/draw.tsx:
```diff
@@ -319,20 +319,252 @@ export class DRAW_Text extends DRAW_Base {
+    const newHash = uuid();
+    drawingFunction.hash = () => newHash;
+
+    const bounds = DRAW_Get_Bounds.getDrawingBounds(
```
wondering, instead of setting the drawingfunction hash, you can just send in this UUID here into the getbrounds "callchain", a little cleaner.

fakob replied:
The getDrawingBounds function uses the hash for lookup and not the callchain. I am not fully sure how your logic works, but I believe it did not update when I passed in the UUID as callchain.

magnificus commented on src/nodes/draw/draw.tsx:
```diff
@@ -319,20 +319,252 @@ export class DRAW_Text extends DRAW_Base {
   }
 }
 
+const bgColorName = 'Background color';
```
Because of how large and complex this is, can we move into a separate file?

fakob replied:
Good idea. Will do.

magnificus commented on src/nodes/draw/draw.tsx:
```diff
@@ -356,30 +589,220 @@ export class DRAW_Combine extends DRAW_Base {
+    const hasMarginOrBackground =
+      inputObject[paddingSocketName] > 0 || bgColor.alpha() > 0;
+
+    if (hasMarginOrBackground) {
```
I will approve PR, but because of how complex this is I likely wont be able to bugfix it if it would break

magnificus commented on src/pixi/utils-pixi.tsx:
```diff
@@ -106,10 +106,17 @@ export const getNodesUpperLeftCorner = (
   let bounds = new PIXI.Rectangle();
   nodes.forEach((node: PIXI.DisplayObject, index: number) => {
-    const tempRect = node.getLocalBounds();
+    console.log(node.children[0].children[0]);
```
I would not use this function this way it looks quite messy, if you want to do this I recommend just changing it in utlils where it is called instead of making this function less predictable.

fakob replied:
Ok, can do.
